blockchain/src/main/scala/co/topl/blockchain/MempoolBroadcaster.scala

```
@@ -8,20 +9,22 @@ import fs2.concurrent.Topic
 
 object MempoolBroadcaster {
 
-  def make[F[_]: Async](
-    mempool:           MempoolAlgebra[F],
-    txsAdoptionsTopic: Topic[F, TypedIdentifier]
-  ): F[MempoolAlgebra[F]] =
-    Async[F].delay {
-      new MempoolAlgebra[F] {
-        def read(blockId: TypedIdentifier): F[Set[TypedIdentifier]] = mempool.read(blockId)
+  def make[F[_]: Async](mempool: MempoolAlgebra[F]): Resource[F, (MempoolAlgebra[F], Topic[F, TypedIdentifier])] =
+    Resource
+      .make(Topic[F, TypedIdentifier])(_.close.void)
+      .map { txsAdoptionsTopic =>
+        val interpreter =
+          new MempoolAlgebra[F] {
+            def read(blockId: TypedIdentifier): F[Set[TypedIdentifier]] = mempool.read(blockId)
 
-        def add(transactionId: TypedIdentifier): F[Unit] =
-          mempool.add(transactionId) >>
-          txsAdoptionsTopic.publish1(transactionId).map(_.leftMap(_ => ())).map(_.merge)
+            def add(transactionId: TypedIdentifier): F[Unit] =
+              mempool.add(transactionId) >>
+              txsAdoptionsTopic.publish1(transactionId).map(_.leftMap(_ => ())).map(_.merge)
```

should we raise the exception in case topic is close, like the local chain broadcaster

---

:+1: Good idea